README.md
Flask-Pusher
============
[![Build Status](https://travis-ci.org/iurisilvio/Flask-Pusher.svg?branch=master)](https://travis-ci.org/iurisilvio/Flask-Pusher)
[![Coverage Status](https://coveralls.io/repos/iurisilvio/Flask-Pusher/badge.png?branch=master)](https://coveralls.io/r/iurisilvio/Flask-Pusher?branch=master)


Flask extension for Pusher. It is a thin wrapper around the official client,
binding Flask app to Pusher client.

Installation
------------

Install `Flask-Pusher` module from PyPI.

```
pip install Flask-Pusher
```

Configuration
-------------

The basic configuration for Pusher is `app_id`, `key` and `secret`. These values are available in Pusher web interface.

```python
PUSHER_APP_ID = 'your-pusher-app-id'
PUSHER_KEY = 'your-pusher-key'
PUSHER_SECRET = 'your-pusher-secret'
```

You can connect to a custom host/port, with these configurations:

```python
PUSHER_HOST = 'api.pusherapp.com'
PUSHER_PORT = '80'
```

The extension auto configure the Pusher encoder to use the `app.json_encoder`.

Usage
-----

This extension simplify Pusher configuration and bind the client to your app.

```python
from flask import Flask
from flask_pusher import Pusher

app = Flask(__name__)
pusher = Pusher(app)
# Use pusher = Pusher(app, url_prefix="/yourpath") to mount the plugin in another path
```

The extension gives you two ways to access the pusher client:

```python
# you can just get the client from current_app
client = current_app.extensions["pusher"]

# or you can get it from the extension
client = pusher.client
```

In both cases, it is a reference to the pusher client.

```
client.trigger('channel_name', 'event', {
    'message': msg,
})
```

Check the docs for the Pusher python client here: http://pusher.com/docs/server_api_guide#/lang=python

Pusher authentication
---------------------

Pusher has authenticated private and presence channels. `Flask-Pusher` create
the `/pusher/auth` route to handle it. To support these authenticated routes,
just decorate a function with `@pusher.auth`.

This function must return `True` for authorized and `False` for unauthorized
users. It happens in the request context, so you have all `Flask` features,
including for example the `Flask-Login` current user.

Set the `PUSHER_AUTH` configuration to change the auth endpoint. The default value is `/auth`.

```python
from flask_login import current_user

@pusher.auth
def pusher_auth(channel_name, socket_id):
    if 'foo' in channel_name:
        # refuse foo channels
        return False
    # authorize only authenticated users
    return current_user.is_authenticated()
```

It also transparently supports batch auth, based on `pusher-js-auth`: https://github.com/dirkbonhomme/pusher-js-auth`. The authentication function is called for each channel in the batch.

Read more about user authentication here: http://pusher.com/docs/authenticating_users


Pusher channel data
-------------------

Presence channels require `channel_data`. `Flask-Pusher` send by default the
`user_id` with the `socket_id`, because it is a required field.

The `@pusher.channel_data` gives you a way to set other values. If a `user_id`
key is returned, it overrides the default `user_id`.

```python
from flask_login import current_user

@pusher.channel_data
def pusher_channel_data(channel_name, socket_id):
    return {
        "name": current_user.name
    }
```

Pusher webhooks
---------------

Pusher has webhooks to send websocket events to your server.

Flask-Pusher create the routes to handle these webhooks and validate the headers `X-Pusher-Key` and `X-Pusher-Signature`.

```python
from flask import request

@pusher.webhooks.channel_existence
def channel_existence_webhook():
    print request.json

@pusher.webhooks.presence
def presence_webhook():
    print request.json

@pusher.webhooks.client
def client_webhook():
    print request.json
```

The JSON request is documented in Pusher docs: http://pusher.com/docs/webhooks

These webhooks routes are mounted in `/pusher/events/channel_existence`, `/pusher/events/presence` and `/pusher/events/client`. Configure your Pusher app to send webhooks to these routes.


Disclaimer
----------
This project is not affiliated with Pusher or Flask.


CHANGELOG.md
Changelog
=========
3.0
 * Drop Pusher<1.7 support
 * Drop Flask<0.12 support
 * Support latest Flask (1.1.1) and Pusher (2.1.14)

2.0.2
 * Package README, LICENSE and CHANGELOG

2.0.1
 * Fix README

2.0
 * Remove Python 2.6 and 3.3 support
 * Support `pusher<3`

1.2
 * Support `pusher<2.0`

1.1.1
 * Remove deprecated `flask.ext.*` calls

1.1
 * Support `pusher` 1.2+
 * Support `PUSHER_SSL` configuration

1.0.1
 * Monkey patch `pusher.json` module to use `flask.json`

1.0
 * `pusher>=1.0` support
 * Custom `JSONEncoder` does not work with `pusher==1.0`

0.4.1
 * Fix dependency version: `pusher<1.0`

0.4
 * Batch auth support
 * Configurable url prefix and authentication path

0.3
 * Webhooks support
 * JSONP authentication support with Flask-Jsonpify

0.2
---
 * Authentication support

0.1
---
 * Initial version
 * Flask app binding to pusher client


tests.py
import unittest
try:
    from unittest import mock
except ImportError:
    import mock
from decimal import Decimal

import pusher as _pusher
from flask import Flask, json, render_template_string, url_for
from flask_pusher import Pusher

pusher_conf = {
    "PUSHER_APP_ID": "1234",
    "PUSHER_KEY": "KEY",
    "PUSHER_SECRET": "SUPERSECRET",
}

SOCKET_ID = "1.42"


class CustomJSONEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, Decimal):
            return str(o)
        return super(CustomJSONEncoder, self).default(o)


class PusherClientTest(unittest.TestCase):

    def setUp(self):
        self.app = Flask(__name__)
        self.app.config.update(pusher_conf)

    def test_default_config(self):
        # fallback to Pusher globals still works
        pusher = Pusher(self.app)
        with self.app.test_request_context():
            self.assertIsNotNone(pusher.client)

    def test_lazy_init_app(self):
        pusher = Pusher()
        pusher.init_app(self.app)
        with self.app.test_request_context():
            self.assertIsNotNone(pusher.client)

    def test_create_extensions_map(self):
        del self.app.extensions
        pusher = Pusher(self.app)
        with self.app.test_request_context():
            self.assertIsNotNone(pusher.client)

    def test_json_encoder(self):
        self.app.json_encoder = CustomJSONEncoder
        pusher = Pusher(self.app)

        with self.app.test_request_context():
            try:
                enc = pusher.client._json_encoder
            except AttributeError:
                enc = pusher.client.encoder
        self.assertEqual(CustomJSONEncoder, enc)

    def test_configuration(self):
        self.app.config["PUSHER_HOST"] = "example.com"
        self.app.config["PUSHER_PORT"] = 8080
        pusher = Pusher(self.app)
        with self.app.test_request_context():
            self.assertIsNotNone(pusher.client)
            self.assertEqual("KEY", pusher.client.key)
            self.assertEqual("SUPERSECRET", pusher.client.secret)
            self.assertEqual("example.com", pusher.client.host)
            self.assertEqual(8080, pusher.client.port)

    def test_all_configurations(self):
        backend = mock.Mock()

        self.app.config.update({
            "PUSHER_SSL": False,
            "PUSHER_TIMEOUT": 3,
            "PUSHER_CLUSTER": "eu",
            "PUSHER_BACKEND": backend,
            "PUSHER_BACKEND_OPTIONS": {"anything": True},
            "PUSHER_NOTIFICATION_HOST": "example.com",
            "PUSHER_NOTIFICATION_SSL": True,
            "PUSHER_ENCRYPTION_MASTER_KEY": "SUPERSECRET",
        })
        pusher = Pusher(self.app)
        with self.app.test_request_context():
            self.assertIsNotNone(pusher.client)
            self.assertFalse(pusher.client.ssl)
            self.assertEqual(3, pusher.client.timeout)
            self.assertEqual("api-eu.pusher.com", pusher.client.host)

            try:
                self.assertEqual(b"SUPERSECRET", pusher.client._encryption_master_key)
            except AttributeError:
                pass

            self.assertTrue(backend.called)
            self.assertTrue(backend.call_args[1]["anything"])

    def test_pusher_key_in_template(self):
        Pusher(self.app)
        with self.app.test_request_context():
            rendered = render_template_string("{{ PUSHER_KEY }}")
            self.assertEqual("KEY", rendered)


class PusherAuthTest(unittest.TestCase):

    def setUp(self):
        self.app = Flask(__name__)
        self.app.debug = True
        self.app.config.update(pusher_conf)
        self.pusher = Pusher(self.app)
        self.client = self.app.test_client()

    def test_url_for(self):
        with self.app.test_request_context():
            url = url_for("pusher.auth")
        self.assertEqual("/pusher/auth", url)

    def test_forbidden_withuot_auth_handler(self):
        response = self.client.post("/pusher/auth")
        self.assertEqual(403, response.status_code)

    def test_auth_refused(self):
        self.pusher.auth(lambda c, s: False)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "private-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(403, response.status_code)

    def test_auth_accepted(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "private-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertIn("auth", data)
        self.assertNotIn("channel_data", data)

    def test_no_channel_data_in_private_channel(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "private-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertIn("auth", data)
        self.assertNotIn("channel_data", data)

    def test_default_channel_data_in_presence_channel(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "presence-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertIn("auth", data)
        channel_data = json.loads(data["channel_data"])
        self.assertEqual({"user_id": SOCKET_ID}, channel_data)

    def test_channel_data_in_presence_channel(self):
        self.pusher.auth(lambda c, s: True)
        self.pusher.channel_data(lambda c, s: {"foo": "bar"})
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "presence-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertIn("auth", data)
        channel_data = json.loads(data["channel_data"])
        self.assertEqual(SOCKET_ID, channel_data["user_id"])
        self.assertIn("bar", channel_data["foo"])

    def test_invalid_channel(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name": "foo",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(404, response.status_code)

    def test_pretty_print_true(self):
        self.app.config['JSONIFY_PRETTYPRINT_REGULAR'] = True
        self.test_no_channel_data_in_private_channel()


class PusherBatchAuthTest(unittest.TestCase):

    def setUp(self):
        self.app = Flask(__name__)
        self.app.debug = True
        self.app.config.update(pusher_conf)
        self.pusher = Pusher(self.app)
        self.client = self.app.test_client()

    def test_one_channel(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name[0]": "private-a",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertEqual(1, len(data))

        data_a = data.get("private-a")
        self.assertEqual(200, data_a["status"])
        self.assertTrue(data_a["data"])

    def test_more_channels(self):
        self.pusher.auth(lambda c, s: "b" not in c)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name[0]": "private-a",
                                          "channel_name[1]": "private-b",
                                          "channel_name[2]": "presence-c",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(200, response.status_code)
        data = json.loads(response.data)
        self.assertEqual(3, len(data))

        a = data.get("private-a")
        self.assertEqual(200, a["status"])
        self.assertIn("auth", a["data"])

        b = data.get("private-b")
        self.assertEqual(403, b["status"])

        c = data.get("presence-c")
        self.assertEqual(200, c["status"])
        c_data = c["data"]
        self.assertIn("auth", c_data)
        self.assertIn("channel_data", c_data)

    def test_missing_channel(self):
        self.pusher.auth(lambda c, s: True)
        response = self.client.post("/pusher/auth",
                                    data={"channel_name[1]": "private-b",
                                          "socket_id": SOCKET_ID})
        self.assertEqual(400, response.status_code)


class PusherWebhookTest(unittest.TestCase):
    def setUp(self):
        self.app = Flask(__name__)
        self.app.debug = True
        self.app.config.update(pusher_conf)
        self.pusher = Pusher()
        self.client = self.app.test_client()
        self._called = False

        @self.pusher.webhooks.client
        def c():
            self._called = True

        self.pusher.init_app(self.app)

    def test_no_webhook(self):
        with self.app.test_request_context():
            url = url_for("pusher.presence_event")
        response = self.client.post(url)
        self.assertEqual(404, response.status_code)
        self.assertFalse(self._called)

    def test_without_key_forbidden(self):
        with self.app.test_request_context():
            url = url_for("pusher.client_event")
        response = self.client.post(url)
        self.assertEqual(403, response.status_code)
        self.assertFalse(self._called)

    def test_invalid_key_forbidden(self):
        with self.app.test_request_context():
            url = url_for("pusher.client_event")
        response = self.client.post(url, headers={
            "Content-Type": "application/json",
            "X-Pusher-Key": "meh"
        })
        self.assertEqual(403, response.status_code)
        self.assertFalse(self._called)

    def test_valid_key_forbidden_without_signature(self):
        with self.app.test_request_context():
            url = url_for("pusher.client_event")
        response = self.client.post(url, headers={
            "Content-Type": "application/json",
            "X-Pusher-Key": "KEY"
        })
        self.assertEqual(403, response.status_code)
        self.assertFalse(self._called)

    def test_invalid_signature(self):
        with self.app.test_request_context():
            url = url_for("pusher.client_event")
        response = self.client.post(url, headers={
            "Content-Type": "application/json",
            "X-Pusher-Key": "KEY",
            "X-Pusher-Signature": "x"
        })
        self.assertEqual(403, response.status_code)
        self.assertFalse(self._called)

    def test_valid_signature(self):
        data = '{"a": "b"}'
        with self.app.test_request_context():
            url = url_for("pusher.client_event")
            signature = self.pusher._sign(data)
        response = self.client.post(url, data=data, headers={
            "Content-Type": "application/json",
            "X-Pusher-Key": "KEY",
            "X-Pusher-Signature": signature
        })
        self.assertEqual(200, response.status_code)
        self.assertTrue(self._called)

    def test_hook_all_handlers(self):
        @self.pusher.webhooks.presence
        def h1():
            pass

        @self.pusher.webhooks.channel_existence
        def h2():
            pass


if __name__ == '__main__':
    unittest.main()


flask_pusher.py
from flask import Blueprint, current_app, request, abort, json
import pusher as _pusher
from pusher.signature import sign, verify

try:
    import flask_jsonpify
except ImportError:  # pragma: no cover
    from flask import jsonify
else:
    jsonify = flask_jsonpify.jsonify
    flask_jsonpify__dumps = flask_jsonpify.__dumps

    def __dumps(*args, **kwargs):
        indent = None
        if current_app.config.get('JSONIFY_PRETTYPRINT_REGULAR', False):
            indent = 2

        return json.dumps(
            args[0] if len(args) is 1 else dict(*args, **kwargs),
            indent=indent
        )

    flask_jsonpify.__dumps = __dumps


class _Pusher(_pusher.Pusher):
    """
    Pusher client wrapper to get attributes from `_pusher_client`
    if the attribute does not exist.

    Provide backward compatibility to `pusher>=1.6`.
    """
    def __getattr__(self, attr):
        client = self._pusher_client
        return getattr(client, attr)


class Pusher(object):

    def __init__(self, app=None, url_prefix="/pusher"):
        self.app = app
        self._auth_handler = None
        self._channel_data_handler = None
        self._blueprint = Blueprint('pusher', __name__, url_prefix=url_prefix)
        self.webhooks = Webhooks(self)

        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        # if config not defined, Pusher will fallback to default config
        app.config.setdefault("PUSHER_APP_ID", '')
        app.config.setdefault("PUSHER_KEY", '')
        app.config.setdefault("PUSHER_SECRET", '')
        app.config.setdefault("PUSHER_HOST", '')
        app.config.setdefault("PUSHER_PORT", '')
        app.config.setdefault("PUSHER_AUTH", '/auth')

        pusher_kwargs = dict(
            app_id=app.config["PUSHER_APP_ID"],
            key=app.config["PUSHER_KEY"],
            secret=app.config["PUSHER_SECRET"],
            host=app.config["PUSHER_HOST"],
            port=app.config["PUSHER_PORT"],
        )

        ssl = app.config.get('PUSHER_SSL')
        if ssl is not None:
            pusher_kwargs["ssl"] = ssl

        timeout = app.config.get('PUSHER_TIMEOUT')
        if timeout is not None:
            pusher_kwargs["timeout"] = timeout

        cluster = app.config.get('PUSHER_CLUSTER')
        if cluster is not None:
            pusher_kwargs["cluster"] = cluster

        backend = app.config.get('PUSHER_BACKEND')
        if backend is not None:
            pusher_kwargs["backend"] = backend

        notification_host = app.config.get('PUSHER_NOTIFICATION_HOST')
        if notification_host is not None:
            pusher_kwargs["notification_host"] = notification_host

        notification_ssl = app.config.get('PUSHER_NOTIFICATION_SSL')
        if notification_ssl is not None:
            pusher_kwargs["notification_ssl"] = notification_ssl

        encryption_master_key = app.config.get('PUSHER_ENCRYPTION_MASTER_KEY')
        if encryption_master_key is not None:
            pusher_kwargs["encryption_master_key"] = encryption_master_key

        pusher_kwargs.update({
            "json_encoder": getattr(app, "json_encoder", None),
            "json_decoder": getattr(app, "json_decoder", None),
        })

        backend_options = app.config.get('PUSHER_BACKEND_OPTIONS')
        if backend_options is not None:
            pusher_kwargs.update(backend_options)

        client = _Pusher(**pusher_kwargs)

        self._make_blueprint(app.config["PUSHER_AUTH"])
        app.register_blueprint(self._blueprint)

        if not hasattr(app, "extensions"):
            app.extensions = {}
        app.extensions["pusher"] = client

    @property
    def client(self):
        return current_app.extensions.get("pusher")

    def auth(self, handler):
        self._auth_handler = handler
        return handler

    def channel_data(self, handler):
        self._channel_data_handler = handler
        return handler

    def _make_blueprint(self, auth_path):
        bp = self._blueprint

        @bp.route(auth_path, methods=["POST"])
        def auth():
            if not self._auth_handler:
                abort(403)

            socket_id = request.form["socket_id"]
            channel_name = request.form.get("channel_name")
            if channel_name:
                response = self._auth_simple(socket_id, channel_name)
                if not response:
                    abort(403)
            else:
                response = self._auth_buffered(socket_id)
            return jsonify(response)

        @bp.app_context_processor
        def pusher_data():
            return {
                "PUSHER_KEY": self.client.key
            }

    def _sign(self, message):
        return sign(self.client.secret, message)

    def _verify(self, message, signature):
        if not signature:
            return False
        return verify(self.client.secret, message, signature)

    def _auth_simple(self, socket_id, channel_name):
        if not self._auth_handler(channel_name, socket_id):
            return None
        return self._auth_key(socket_id, channel_name)

    def _auth_buffered(self, socket_id):
        response = {}
        while True:
            n = len(response)
            channel_name = request.form.get("channel_name[%d]" % n)
            if not channel_name:
                if n == 0:
                    # it is not a buffered request
                    abort(400)
                break
            r = {}
            auth = self._auth_simple(socket_id, channel_name)
            if auth:
                r.update(status=200, data=auth)
            else:
                r.update(status=403)
            response[channel_name] = r
        return response

    def _auth_key(self, socket_id, channel_name):
        if channel_name.startswith("presence-"):
            channel_data = {"user_id": socket_id}
            if self._channel_data_handler:
                d = self._channel_data_handler(channel_name, socket_id)
                channel_data.update(d)
            auth_args = [socket_id, channel_data]
        elif channel_name.startswith("private-"):
            auth_args = [socket_id]
        else:
            # must never happen, this request is not from pusher
            abort(404)

        return self.client.authenticate(channel_name, *auth_args)


class Webhooks(object):
    CHANNEL_EXISTENCE_EVENT = "channel_existence"
    PRESENCE_EVENT = "presence"
    CLIENT_EVENT = "client"

    def __init__(self, pusher):
        self.pusher = pusher
        self._handlers = {}
        self._register(self.CHANNEL_EXISTENCE_EVENT)
        self._register(self.PRESENCE_EVENT)
        self._register(self.CLIENT_EVENT)

    def channel_existence(self, func):
        self._handlers[self.CHANNEL_EXISTENCE_EVENT] = func
        return func

    def presence(self, func):
        self._handlers[self.PRESENCE_EVENT] = func
        return func

    def client(self, func):
        self._handlers[self.CLIENT_EVENT] = func
        return func

    def _register(self, event):
        def route():
            func = self._handlers.get(event)
            if not func:
                abort(404)
            self._validate()
            func()
            return "OK", 200

        rule = "/events/%s" % event
        name = "%s_event" % event
        self.pusher._blueprint.add_url_rule(rule, name, route,
                                            methods=["POST"])

    def _validate(self):
        pusher_key = request.headers.get("X-Pusher-Key")
        if pusher_key != self.pusher.client.key:
            # invalid pusher key
            abort(403)

        webhook_signature = request.headers.get("X-Pusher-Signature")
        if not self.pusher._verify(request.data.decode(), webhook_signature):
            # invalid signature
            abort(403)
